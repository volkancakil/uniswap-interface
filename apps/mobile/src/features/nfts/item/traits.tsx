import dayjs from 'dayjs'
import React from 'react'
import { ListRenderItemInfo, StyleSheet } from 'react-native'
import { FlatList } from 'react-native-gesture-handler'
import { Flex, Text, useSporeColors } from 'ui/src'
import { spacing } from 'ui/src/theme'
import { NftAssetTrait } from 'uniswap/src/data/graphql/uniswap-data-api/__generated__/types-and-hooks'

const formatTraitValue = (trait: NftAssetTrait): string | undefined => {
  if (!trait.value) {
    return undefined
  }

  if (trait.name?.toLowerCase().split(' ').includes('date')) {
    const date = dayjs(+trait.value * 1000)
    return date.isValid() ? date.format('MMM D, YYYY') : trait.value
  }

  return trait.value
}

export function NFTTraitCard({
  trait,
  titleTextColor,
}: {
  trait: NftAssetTrait
  titleTextColor?: string
}): JSX.Element {
  const colors = useSporeColors()

  return (
    <Flex
      backgroundColor={colors.surface2.val}
      borderRadius="$rounded16"
      gap="$spacing4"
      px="$spacing16"
      py="$spacing12"
    >
      <Text style={{ color: titleTextColor ?? colors.neutral2.get() }} variant="body3">
        {trait.name && trait.name.charAt(0).toUpperCase() + trait.name.slice(1).toLowerCase()}
      </Text>
      <Text color={colors.neutral1.val} variant="subheading2">
        {formatTraitValue(trait)}
      </Text>
    </Flex>
  )
}

export function NFTTraitList({
  traits,
  titleTextColor = 'neutral1',
}: {
  traits: NftAssetTrait[]
  titleTextColor?: string
}): JSX.Element {
  function renderItem(item: ListRenderItemInfo<NftAssetTrait>): JSX.Element {
    return <NFTTraitCard titleTextColor={titleTextColor} trait={item.item} />
  }

  return (
    <FlatList
      ItemSeparatorComponent={Separator}
      contentContainerStyle={Styles.listContainer}
      data={traits}
      horizontal={true}
      renderItem={renderItem}
      showsHorizontalScrollIndicator={false}
    />
  )
}

function Separator(): JSX.Element {
  return <Flex width={spacing.spacing8} />
}

const Styles = StyleSheet.create({
  listContainer: {
    paddingHorizontal: spacing.spacing24,
  },
})
