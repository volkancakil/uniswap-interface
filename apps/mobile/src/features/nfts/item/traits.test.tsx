import React from 'react'
import { NFTTraitCard } from 'src/features/nfts/item/traits'
import { render } from 'src/test/test-utils'
import { NFT_ASSET_TRAIT } from 'uniswap/src/test/fixtures'

it('renders trait card', () => {
  const tree = render(<NFTTraitCard trait={NFT_ASSET_TRAIT} />)
  expect(tree).toMatchSnapshot()
})
