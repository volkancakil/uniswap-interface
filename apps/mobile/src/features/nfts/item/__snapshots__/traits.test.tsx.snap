// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders trait card 1`] = `
<View
  style={
    {
      "backgroundColor": "#F9F9F9",
      "borderBottomLeftRadius": 16,
      "borderBottomRightRadius": 16,
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "flexDirection": "column",
      "gap": 4,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      {
        "color": {
          "dynamic": {
            "dark": "#9B9B9B",
            "light": "#7D7D7D",
          },
        },
        "fontFamily": "Basel Grotesk",
        "fontSize": 15,
        "fontWeight": "400",
        "lineHeight": 20,
      }
    }
    suppressHighlighting={true}
  >
    Traitname
  </Text>
  <Text
    allowFontScaling={true}
    maxFontSizeMultiplier={1.4}
    style={
      {
        "color": "#222222",
        "fontFamily": "Basel Grotesk",
        "fontSize": 17,
        "fontWeight": "400",
        "lineHeight": 24,
      }
    }
    suppressHighlighting={true}
  >
    traitValue
  </Text>
</View>
`;
